src/rdfTypes/GeometryInfoHelpersImpl.h

```diff
@@ -190,6 +209,139 @@ inline util::geo::DBox projectInt32WebMercToDoubleLatLng(
           projectInt32WebMercToDoubleLatLng(box.getUpperRight())};
 };
 
+// TODO
+constexpr double EARTH_RADIUS_METERS = 6'371'000.0;
+constexpr double STERADIAN_TO_M2 = EARTH_RADIUS_METERS * EARTH_RADIUS_METERS;
+
+// Helper to convert a libspatialjoin `Ring` to an `S2Loop`
+inline std::unique_ptr<S2Loop> makeS2Loop(const Ring<CoordType>& ring) {
+  std::vector<S2Point> points;
+  for (const auto& latlon : ring) {
+    S2LatLng s2latlng = S2LatLng::FromDegrees(latlon.getY(), latlon.getX());
+    points.push_back(s2latlng.ToPoint());
+  }
+
+  // Ensure that there are no zero-length edges (that is edges with twice the
+  // same point), as this will lead to an exception from `S2Loop`.
+  std::vector<S2Point> cleaned;
+  for (size_t i = 0; i < points.size(); ++i) {
+    if (i == 0 || points.at(i) != points.at(i - 1)) {
+      cleaned.push_back(points.at(i));
+    }
+  }
+  if (cleaned.front() == cleaned.back()) {
+    cleaned.pop_back();
+  }
+
+  auto loop = std::make_unique<S2Loop>(cleaned);
+  loop->Normalize();
+  if (!loop->IsValid()) {
+    throw InvalidPolygonError();
+  }
+  return loop;
+}
+
+inline S2Polygon makeS2Polygon(const Polygon<CoordType>& polygon) {
+  std::vector<std::unique_ptr<S2Loop>> loops;
+
+  // Outer boundary
+  loops.push_back(makeS2Loop(polygon.getOuter()));
+
+  // Holes
+  for (const auto& hole : polygon.getInners()) {
+    loops.push_back(makeS2Loop(hole));
+  }
+
+  S2Polygon s2polygon;
+  s2polygon.InitNested(std::move(loops));
+  return s2polygon;
+}
+
+inline double computeMetricAreaS2Polygon(const S2Polygon& polygon) {
+  return polygon.GetArea() * STERADIAN_TO_M2;
+}
+
+// Compute the area of a polygon in square meters on earth using s2
+inline double computeMetricAreaPolygon(const Polygon<CoordType>& polygon) {
+  return computeMetricAreaS2Polygon(makeS2Polygon(polygon));
+}
+
+// Compute the area of a multipolygon in square meters on earth using s2
+inline double computeMetricAreaMultiPolygon(
+    const MultiPolygon<CoordType>& polygons) {
+  // Empty multipolygon has empty area
+  if (polygons.empty()) {
+    return 0.0;
+  }
+  // Multipolygon with one member has exactly area of this member
+  if (polygons.size() == 1) {
+    return computeMetricAreaPolygon(polygons.at(0));
+  }
+  // For a multipolygon with multiple members, we need to compute the union of
+  // the polygons to determine their area.
+  // TODO<ullingerc>: the number of steps could be reduced by building the union
+  // in pairs (like divide-and-conquer) or using `S2Builder`
+  auto unionPolygon = makeS2Polygon(polygons.at(0));
+  for (size_t i = 1; i < polygons.size(); ++i) {
+    unionPolygon.InitToUnion(unionPolygon, makeS2Polygon(polygons.at(i)));
+  }
+  return computeMetricAreaS2Polygon(unionPolygon);
+};
```

Comment:
Does this have a reasonable behavior for all the inner/outer etc behavior that is preserved when being converted to S2? do you test for this?